src/main/java/io/confluent/kafkarest/ProducerPool.java

```diff
@@ -186,16 +199,57 @@ private static String getBootstrapBrokers(ZkUtils zkUtils) {
       String topic,
       Integer partition,
       EmbeddedFormat recordFormat,
+      SecurityContext securityContext,
       SchemaHolder schemaHolder,
       Collection<? extends ProduceRecord<K, V>> records,
       ProduceRequestCallback callback
   ) {
     ProduceTask task = new ProduceTask(schemaHolder, records.size(), callback);
     log.trace("Starting produce task " + task.toString());
-    RestProducer restProducer = producers.get(recordFormat);
+    RestProducer restProducer = getRestProducer(recordFormat, getJaasModule(securityContext));
     restProducer.produce(task, topic, partition, records);
   }
 
+  private RestProducer getRestProducer(EmbeddedFormat recordFormat, JaasModule jaasModule) {
+    RestProducer restProducer;
+    if (jaasModule != null) {
+      if (!authenticatedProducers.containsKey(jaasModule)) {
+        restProducer = buildAuthenticatedProducer(recordFormat, jaasModule);
+        authenticatedProducers.put(jaasModule, restProducer);
```

Comment:
True, do you have an idea to improve it ? 

I could create the producer "on demand" to reduce memory usage but producing messages will be slower since we will need to create a producer for each request.

> each user principal will have its own producer, if we have many principals, like in our case we have 20/25 principals, we will not get a good throughput from the producer if they all are producing to same topic.

As a workaround you could use the same principal to produce messages on a topic. For instance, user1, user2 and user3 are allowed to produce messages on topic A. If user1, user2 or user3 is authenticated then you could return a "generic" principal to produce messages on topic A.